### PR TITLE
coverage: try use a unified idea to increase each stack if enabled

### LINKED ISSUE #44062: Need a way to deal with stack size needed when running coverage report.
**Describe the bug**
This is an extension of [#41388](https://github.com/zephyrproject-rtos/zephyr/issues/41388), when we run the code coverage report, some of the tests will fail due to not have sufficient stack size.
(Ex. tests/net, it needs more stack size like that to pass the test: 
CONFIG_MAIN_STACK_SIZE=4096
CONFIG_NET_TCP_WORKQ_STACK_SIZE=2048
CONFIG_SYSTEM_WORKQUEUE_STACK_SIZE=4096

Just enlarge the stack size or extra coverage.conf to decide stack size needed seems not a good solution for this issue.
So a better way is needed to make it won't need to add extra stack sizes manually when running with coverage report generated.

The point of this requirement might be: We don't need to add extra stack size case by case, and boards.

### PR DESCRIPTION
**Pull request by Zhaoningx**
As we know, if coverage is enabled, each stack will be needed more,
lots of cases are failed to generate coverage reports due to stack
overflow, so I try to use a unified idea to increase each stack.

I have manually tested these cases, and successfully generate
coverage reports for failed test cases by increasing the stack.
And it may be cumbersome to change one by one, so I want
to try to use the macro `CONFIG_COVERAGE_STACK_SIZE` to
expand different stacks at the same time for each test code.
After these days of testing, there are still some issues work in process.
I want to push a draft PR first to see if the idea works.

Fixes #44062

Signed-off-by: NingX Zhao <ningx.zhao@intel.com>

### CODE REVIEW

**Comment by enjiamai** [comment]
Hi @Zhaoningx, could you please rebase it and address the failure in CI, thanks.

**Review by dcpleung — DISMISSED** [review]

**Comment by carlocaione** [comment]
TBH I'm not a huge fan of having one more `_STACK_SIZE` only for tests. We already have `TEST_EXTRA_STACK_SIZE`, can't we simply make that widely used in tests and extend it when coverage is enabled?

File: kernel/init.c [hunk 1]
```diff
@@ -331,8 +331,8 @@ static char *prepare_multithreading(void)
 	_kernel.ready_q.cache = &z_main_thread;
 #endif
 	stack_ptr = z_setup_new_thread(&z_main_thread, z_main_stack,
-				       CONFIG_MAIN_STACK_SIZE, bg_thread_main,
-				       NULL, NULL, NULL,
+				       CONFIG_MAIN_STACK_SIZE + CONFIG_COVERAGE_STACK_SIZE,
+					   bg_thread_main, NULL, NULL, NULL,
```
Comment (stephanosio) [inline kernel/init.c:335]: Nit (style issue): unnecessary indentation change (you seem to be using TAB with the column width of 4; please use 8).

File: arch/arm/core/aarch32/mpu/arm_core_mpu.c [hunk 2]
```diff
@@ -46,7 +46,7 @@ LOG_MODULE_REGISTER(mpu);
 		_MPU_DYNAMIC_REGIONS_AREA_START)
 
 #if !defined(CONFIG_MULTITHREADING) && defined(CONFIG_MPU_STACK_GUARD)
-K_THREAD_STACK_DECLARE(z_main_stack, CONFIG_MAIN_STACK_SIZE);
+extern K_THREAD_STACK_DEFINE(z_main_stack, CONFIG_MAIN_STACK_SIZE + CONFIG_COVERAGE_STACK_SIZE);
```
Comment (stephanosio) [inline arch/arm/core/aarch32/mpu/arm_core_mpu.c:49]: This is incorrect. `K_THREAD_STACK_DECLARE` (aka. extern) must be used here.

**Review by stephanosio — CHANGES_REQUESTED** [review]

**Comment by Zhaoningx** [comment]
> TBH I'm not a huge fan of having one more `_STACK_SIZE` only for tests. We already have `TEST_EXTRA_STACK_SIZE`, can't we simply make that widely used in tests and extend it when coverage is enabled?

Hi @carlocaione, I also thought about the macro `TEST_EXTRA_STACK_SIZE` to be used for increasing the coverage stack, but as you know, this macro can configure in some test cases prj.conf by the user. If configured, it will overwrite the value that I set in kconfig. So I prefer to use a new macro to do this. Do you think it is OK? And, so far this PR still can't fix all test cases overflow issues due to all test cases needing different stack sizes when coverage is enabled. When I increased the stack very large, some failure test cases that need more stack will be successful, but some test cases maybe build errors or CPU exceptions. Do you think the idea is possible?

**Comment by carlocaione** [comment]
> Hi @carlocaione, I also thought about the macro TEST_EXTRA_STACK_SIZE to be used for increasing the coverage stack, but as you know, this macro can configure in some test cases prj.conf by the user. If configured, it will overwrite the value that I set in kconfig. So I prefer to use a new macro to do this. Do you think it is OK? 

It is ok (that's why I'm not -1 this) but stil I don't like it. The problem that the value can be modified by the users can be true for every single Kconfig symbol, so it's not a problem (the users should be aware of the consequences of their own actions :)

The reason why I don't like this solution is that when a dev will be introducing a new test in the future, they will have to remember to always set the stack size to something like:
```
CONFIG_MAIN_STACK_SIZE + CONFIG_COVERAGE_STACK_SIZE
```
if in the future we will be introducing some other symbol that extends the stack, that will need to be extended to:
```
CONFIG_MAIN_STACK_SIZE + CONFIG_COVERAGE_STACK_SIZE + CONFIG_WHATEVER_STACK_SIZE + ...
```
for all the tests again.

The other solution is having `TEST_EXTRA_STACK_SIZE` already automatically including all the different components so that we can extend that behind the curtains when needed without touching any tests or having to change anything for the user. Not sure how difficult is having that though.

File: arch/arm/core/aarch32/mpu/arm_core_mpu.c (hunk 2 above)
Comment (Zhaoningx) [reply arch/arm/core/aarch32/mpu/arm_core_mpu.c:49]: Thanks for correcting, I'll do

**Comment by Zhaoningx** [comment]
> > Hi @carlocaione, I also thought about the macro TEST_EXTRA_STACK_SIZE to be used for increasing the coverage stack, but as you know, this macro can configure in some test cases prj.conf by the user. If configured, it will overwrite the value that I set in kconfig. So I prefer to use a new macro to do this. Do you think it is OK?
> 
> It is ok (that's why I'm not -1 this) but stil I don't like it. The problem that the value can be modified by the users can be true for every single Kconfig symbol, so it's not a problem (the users should be aware of the consequences of their own actions :)
> 
> The reason why I don't like this solution is that when a dev will be introducing a new test in the future, they will have to remember to always set the stack size to something like:
> 
> ```
> CONFIG_MAIN_STACK_SIZE + CONFIG_COVERAGE_STACK_SIZE
> ```
> 
> if in the future we will be introducing some other symbol that extends the stack, that will need to be extended to:
> 
> ```
> CONFIG_MAIN_STACK_SIZE + CONFIG_COVERAGE_STACK_SIZE + CONFIG_WHATEVER_STACK_SIZE + ...
> ```
> 
> for all the tests again.
> 
> The other solution is having `TEST_EXTRA_STACK_SIZE` already automatically including all the different components so that we can extend that behind the curtains when needed without touching any tests or having to change anything for the user. Not sure how difficult is having that though.

Hi, I agree with you very much, it is very reasonable for users to modify the size of each stack. In the source code, the stack size is different by default when the coverage is enabled and disabled. But if the user configures this size, it will be overwritten directly, regardless of whether the coverage is enabled or not. This is also the main reason why many cases fail when the coverage is enabled. So I defined a macro to increase them separately. Could you detail your solution? I want to have a try and re-push the PR if worked.

**Comment by carlocaione** [comment]
> Could you detail your solution?

No clear idea. Something along these lines maybe: the test framework is emitting a new `ADJUST_TEST_STACK_SIZE(x)` macro that you can use to increase the stack size by a reasonable value taking into account coverage & company. So the user can do:
```
K_THREAD_STACK_DECLARE(z_main_stack, ADJUST_TEST_STACK_SIZE(CONFIG_MAIN_STACK_SIZE));
```
where the `ADJUST_TEST_STACK_SIZE` is something like:
```
#define ADJUST_TEST_STACK_SIZE(x) (x + CONFIG_COVERAGE_STACK_SIZE + CONFIG_WHATEVER + ...)
```

File: arch/arm/core/aarch32/mpu/arm_core_mpu.c (hunk 2 above)
Comment (Zhaoningx) [reply arch/arm/core/aarch32/mpu/arm_core_mpu.c]: Hi sorry for the delay, I've addressed it.

**Comment by Zhaoningx** [comment]
> > Could you detail your solution?
> 
> No clear idea. Something along these lines maybe: the test framework is emitting a new `ADJUST_TEST_STACK_SIZE(x)` macro that you can use to increase the stack size by a reasonable value taking into account coverage & company. So the user can do:
> 
> ```
> K_THREAD_STACK_DECLARE(z_main_stack, ADJUST_TEST_STACK_SIZE(CONFIG_MAIN_STACK_SIZE));
> ```
> 
> where the `ADJUST_TEST_STACK_SIZE` is something like:
> 
> ```
> #define ADJUST_TEST_STACK_SIZE(x) (x + CONFIG_COVERAGE_STACK_SIZE + CONFIG_WHATEVER + ...)
> ```

Hi @carlocaione, Sorry I still don't know how to do it, because now test cases are existing, it seems stacks size needs to be increased from the initialization, it can solve the stack overflow issue. The idea you mentioned is good for users to design new test cases I think.

**Comment by carlocaione** [comment]
Ok, not exactly sure what's the problem with my suggestion but I'm not going to block this. I'm not giving my +1 here either because I don't like this solution though.

**Comment by Zhaoningx** [comment]
Hi @carlocaione, sorry for the delay due to OOO, and I have updated the PR, could you help to review it again? Thanks.

**Comment by carlocaione** [comment]
Sorry, I still do not like this. If, as a user, I'm setting a value for (i.e.) the main stack size, I really would be pissed if someone else would change that under the hood (especially to make it bigger).

Now, this is (arguably?) my personal opinion on this so again I'm not blocking it but I'm not giving my +1 either.

**Comment by Zhaoningx** [comment]
> Sorry, I still do not like this. If, as a user, I'm setting a value for (i.e.) the main stack size, I really would be pissed if someone else would change that under the hood (especially to make it bigger).
> 
> Now, this is (arguably?) my personal opinion on this so again I'm not blocking it but I'm not giving my +1 either.

I totally understand, and I don't change the value when the coverage is disabled. And the main issue is that some stack size values are not enough to generate the coverage report(stack overflow error maybe), you can refer to the issue https://github.com/zephyrproject-rtos/zephyr/issues/44062. I'll have a try to find if there is another solution. Thanks.

**Pull request closed without merging**